### PR TITLE
modified way to patch, can realize params map for compatibility, and try to hook Item in lua.

### PR DESCRIPTION
**Pull request by zhu-rengong**
this PR: modified way to patch, can realize params map for compatibility, and try to hook Item in lua.

I tried to modify the `methodNameToHookName` to `methodPathToHookName`, replacing the method name with the complete method path, so that it would not conflict with other classes.

In order to be compatible with the old Lua script, I did a parameter mapping before calling the hook method, example:
if add hook for Barotrauma.Item.Use at prefix, we just need to do a little work to achieve compatibility.

**Original C# Code**
```c#
public void Use(float deltaTime, Character character = null, Limb targetLimb = null)
{
    ...
    var should = new LuaResult(GameMain.Lua.hook.Call("itemUse", new object[] { this, character, targetLimb }));
}
```

**Add Hook in lua**
```lua
Hook.HookMethod("Barotrauma.Item", "Use", "itemUse", 
    function (instance, deltaTime, character, targetLimb) -- __params
        -- Map __params to the same as the original
        return {instance, character, targetLimb}
    end)
```

**Move Item Hook to Lua**
```lua
Hook.HookMethod("Barotrauma.Item", "TryInteract", "itemInteract", nil) -- instance, picker, ignoreRequiredItems, forceSelectKey, forceActionKey
Hook.HookMethod("Barotrauma.Item", "Use", "itemUse", function (instance, deltaTime, character, targetLimb) return {instance, character, targetLimb} end)
Hook.HookMethod("Barotrauma.Item", "SecondaryUse", "itemSecondaryUse", function (instance, deltaTime, character) return {instance, character} end)
Hook.HookMethod("Barotrauma.Item", "ApplyTreatment", "itemApplyTreatment", nil) -- instance, user, character, targetLimb
Hook.HookMethod("Barotrauma.Item", "Combine", "itemCombine", nil) -- instance, item, user
Hook.HookMethod("Barotrauma.Item", "Drop", "itemDrop", function (instance, dropper, createNetworkEvent)	return {instance, dropper} end)
Hook.HookMethod("Barotrauma.Item", "Equip", "itemEquip", nil) -- instance, character
Hook.HookMethod("Barotrauma.Item", "Unequip", "itemUnequip", nil) -- instance, character
```
Note: The original itemUse and itemSecondaryUse hooks were not called in the method prefix, but I let them do so.

some fix with this commit:
1. Wrong parameter name
2. Replace RegisterType with CreateStatic in DefaultLib.lua

### CODE REVIEW

**Comment by evilfactory** [comment]
Maybe instead of defining hooks directly with `HookMethod`, it should be better to just use a callback and do a hook call inside this callback, like that:

```lua
Hook.HookMethod("Barotrauma.Example", "ExampleFunc", function(a,b,c)
      Hook.Call("exampleHook", a, b, c)
end)
```

**Comment by zhu-rengong** [comment]
1. yeah, it's the better solution
2. warpped parameters for versatility, looks like:
```lua
Hook.HookMethod("Barotrauma.Item", "ApplyTreatment", function (instance, p)
	Hook.Call("itemApplyTreatment", instance, p.user, p.character, p.targetLimb)
end, HookMethodType.Before)

Hook.HookMethod("Barotrauma.Item", "ApplyTreatment", function (instance, p)
	Hook.Call("itemApplyTreatment_after", instance, p.user, p.character, p.targetLimb)
end, HookMethodType.After)
```
3. Don't implement Hooks of Use and SecondlyUse in this way for the time being

File: Barotrauma/BarotraumaShared/SharedSource/Lua/LuaSetup.cs [hunk 1]
```diff
@@ -349,6 +349,7 @@ public void Initialize()
 			lua.Globals["load"] = (Func<string, Table, string, DynValue>)LoadString;
 			
 			lua.Globals["LuaUserData"] = UserData.CreateStatic<LuaUserData>();
+			lua.Globals["HookMethodType"] = UserData.CreateStatic<LuaHook.HookMethodType>();
```
Comment (evilfactory) [inline Barotrauma/BarotraumaShared/SharedSource/Lua/LuaSetup.cs:352]: `HookMethodType` is already accessible by Lua via `Hook.HookMethodType`, so there's no need to create a global static enum of it

Comment (zhu-rengong) [reply Barotrauma/BarotraumaShared/SharedSource/Lua/LuaSetup.cs]: okay, had deleted it.

**Comment by evilfactory** [comment]
Okay, looks good, merging pull request.

**Comment by zhu-rengong** [comment]
In fact, I originally wanted to post it in discussion to share the code, but it didn't seem to work. Because I use git for the first time, I am sorry that the way of using PR has caused a bad influence on your development, and actually the time difference in our area may be twelve hours……

**Comment by evilfactory** [comment]
Don't worry, i'm not very experienced with git too, and sending PRs is fine, it doesn't affect development much, anyway thank you :)

**Pull request merged**